microceph/ceph/services_placement.go

```diff
@@ -0,0 +1,97 @@
+package ceph
+
+import (
+	"fmt"
+
+	"github.com/canonical/lxd/shared/logger"
+	"github.com/canonical/microceph/microceph/api/types"
+	"github.com/canonical/microceph/microceph/common"
+)
+
+type PlacementIntf interface {
+	PopulateParams(common.StateInterface, string) error
+	HospitalityCheck(common.StateInterface) error
+	ServiceInit(common.StateInterface) error
+	PostPlacementCheck(common.StateInterface) error
+	DbUpdate(common.StateInterface) error
+}
+
+func GetServicePlacementTable() map[string](PlacementIntf) {
+	return map[string](PlacementIntf){
+		"mon": &GenericServicePlacement{"mon"},
+		"mgr": &GenericServicePlacement{"mgr"},
+		"mds": &GenericServicePlacement{"mds"},
+		"rgw": &RgwServicePlacement{},
+	}
+}
+
+func ServicePlacementHandler(s common.StateInterface, payload types.EnableService) error {
+	var ok bool
+	var spt = GetServicePlacementTable()
+	var sp PlacementIntf
+
+	logger.Debugf("Enabling %s service, payload: %v", payload.Name, payload.Payload)
+	if sp, ok = spt[payload.Name]; !ok {
+		err := fmt.Errorf("%s enablement is not supported", payload.Name)
+		logger.Error(err.Error())
+		return err
+	}
+
+	if payload.Wait {
+		err := EnableService(s, payload, sp)
+		if err != nil {
+			logger.Errorf("Failed %s service enablement request: %v", payload.Name, err)
+			return err
+		}
+	} else {
+		go func() {
+			// Async call to Enable service.
+			EnableService(s, payload, sp)
```

Comment:
what about error reporting here? Even if it's asynchronous, shouldn't we fetch the error and log it at least?

Reply:
Added outer error log for EnableService in async case. EnableService itself has error logs at each failure.